gsr_booking/booking_logic.py

```diff
@@ -0,0 +1,255 @@
+                    break
+                if member["user__email"] is "":
+                    continue
+
+                # make request to labs-api-server
+                success, error = self.book_room_for_user(
+                    room,
+                    lid,
+                    next_start.strftime(self.DATE_FORMAT_STR),
+                    next_end.strftime(self.DATE_FORMAT_STR),
+                    member["user__email"],
+                )
+                if success:
+                    key = f"{lid}_{room}"
+                    if not key in bookings:
+                        bookings[key] = []
+                    new_bookings = self.split_booking(
+                        next_start, next_end, member["username"], True
+                    )
+                    bookings[key].extend(new_bookings)
+
+                    next_start = next_end
+                    next_end = min(
+                        END_DATE, next_start + datetime.timedelta(hours=self.MAX_SLOT_HRS)
+                    )
+                elif self.is_fatal_error(error):
+                    result_json = self.construct_bookings_json_obj(
+                        bookings, lid, room, END_DATE, next_start, next_end, error
+                    )
+                    return result_json
+                else:
+                    failed_members.append(member)
+
+            # if unbooked slots still remain and not all booking requests succeeded, loop through each member again
+            # but get reservation credits first to see how much we can book
+
+            for member in failed_members:
+                if next_end - next_start < datetime.timedelta(hours=0.1):
+                    print("BOOKED EVERYTHING ALREADY")
+                    break
+                if member["user__email"] is "":
+                    print("empty email address")
+                    continue
+
+                # calculate number of credits already used via getReservations
+                (success, used_credit_hours) = self.get_used_booking_credit_for_user(
+                    lid, member["user__email"]
+                )
+                remaining_credit_hours = self.MAX_SLOT_HRS - used_credit_hours
+                rounded_remaining_credit_hours = math.floor(2 * remaining_credit_hours) / 2
+                if success and remaining_credit_hours >= self.MIN_SLOT_HRS:
+                    next_end = min(
+                        END_DATE,
+                        next_start + datetime.timedelta(hours=rounded_remaining_credit_hours),
+                    )
+                    (success, error) = self.book_room_for_user(
+                        room,
+                        lid,
+                        next_start.strftime(self.DATE_FORMAT_STR),
+                        next_end.strftime(self.DATE_FORMAT_STR),
+                        member["user__email"],
+                    )
+                    if success:
```

Comment:
I think if we keep the blind booking above, this logic here should be factored out into a function to avoid duplication.

Reply:
The problem is I need to update the local variables (`next_start, next_end, bookings`), so if I made a function it would have maybe 1 line (for finding the key). Any other ideas?

Reply:
good point! I think i had a question somewhere in here about what the purpose of a blind booking was? versus just always checking how much availability someone has.

Reply:
For sure - I did respond to that question in another comment but GitHub thinks its "outdated" for some reason, so I'll post my answer here:

The algorithm would be shorter, but slower on average if we skipped the "blind" booking. This is because getting reservations for a user takes ~2x times longer than making a booking for that user. Also, although including "blind" booking does increase the worse case time for booking, in the average case we would only need to get reservation credits for a few users (assuming that some people in the group have not booked anything).

I talked with Josh and he suggested that in the future we can use the booking credit data stored in SQL. Although it only captures the bookings made through pennmobile, we would be able to get reservation credits almost instantly, allowing us to remove the blind booking part and drastically decease time to book for group.